Feature/cubed sphere structured functionspace

odlomax opened the pull request:
I've made a functionspaces `CubedSphereCellCoumns` and `CubedSphereNodeColumns` which add a handful of methods to `CellColumns` and `NodeColumns`. It should make operations like interpolation significantly simpler. `CubedSphereCellCoumns` inherits two classes: `CellCoumns` and `StructuredCubedSphere` (`CubedSphereNodeColumns` does the equivalent).

The new functionspaces have two important methods:

`idx_t index(idx_t t, idx_t i, idx_t j)` which converts an (t, i, j) triplet into a field index.

`void for_each(const functor& f, bool include_halo = false)` which traverses the functionspace, possibly skipping halos, and applies a functor. The functor should have the signature `f(idx_t index, idx_t t, idx_t i, idx_t j)`.

If you're feeling a bit old fashioned, you could also perform a triple loop in (t, j, i) order and skip elements where index(t, i, j) == invalid_index(). However this should be avoided if possible, as it make assumptions on the memory layout of the mesh.

@wdeconinck I realise that the double inheritance is a little unorthodox for Atlas classes Is this okay or do you have a different method you'd prefer?

I've added mainly cosmetic minor changes to the cubed sphere mesh generator and utilities. This refactors "ijt" to "tij" and tweaks some the the interfaces accordingly. I also added more meaningful messages to some of the error traps.

closes #59

odlomax committed:
Branch ready for PR.

odlomax committed:
Made use of struct padding.

odlomax committed:
Made variable names consistent.

odlomax committed:
Corrected global index type.

wdeconinck committed:
Apply clang-format

odlomax committed:
Addressed reviewer comments.

wdeconinck committed:
Apply clang-format

odlomax committed:
Handwritten small matrix operations.

wdeconinck committed:
Fix compilation for older compilers

odlomax committed:
Saving before merge.

odlomax committed:
Merged develop into branch.

odlomax committed:
Refactored ijt to tij.

odlomax committed:
Tidied up ijt to tij refactor.

wdeconinck committed:
Added Github Actions workflow for CI

Co-authored-by: Dusan Vuckovic <dusan@dvuckovic.com>

odlomax committed:
Refactoring utilities.

odlomax committed:
Refactoring utilities.

odlomax committed:
Refactoring utilities.

odlomax committed:
Ready for PR.

odlomax committed:
Ready for PR.

odlomax committed:
Last minute error traps.

odlomax committed:
Dialled backed some of the error traps.

odlomax committed:
Modified some junk in test.

odlomax committed:
Merged base branch.

wdeconinck commented on src/atlas/functionspace/detail/StructuredCubedSphere.h:
```diff
@@ -0,0 +1,113 @@
+  /// j lower bound for tile t (including halo)
+  idx_t j_end(idx_t t) const;
+
+  /// Return array_view index for (i, j, t).
```
```suggestion
  /// Return array_view index for (t, i, j).
```

wdeconinck commented on src/atlas/functionspace/detail/StructuredCubedSphere.h:
```diff
@@ -0,0 +1,113 @@
+  /// Apply functor f(index, t, i, j) to all valid (t, i, j) triplets.
+  /// (use "#pragma omp atomic" to update array elements other than [index].)
+  template<typename functor>
+  void for_each(const functor& f, bool include_halo = false)  const {
```
Perhaps we could somehow agree on a standard syntax `parallel_for`, similar to 
https://github.com/ecmwf/atlas/blob/develop/src/atlas/functionspace/StructuredColumns.h#L220-L227
The first (optional) argument is a "Config" object. It is first so that a lambda functor can then be the final argument.

odlomax replied:
No problem. I think I should be able to implement something similar to your `For` class, and with `FunctorArgs(index, t, i, j, k)` and `FunctorArgs(index, t, i, j)`.

wdeconinck commented on src/atlas/functionspace/detail/StructuredCubedSphere.h:
```diff
@@ -0,0 +1,113 @@
+    // Loop over elements.
+    atlas_omp_parallel_for (idx_t index = 0; index < ijtView_.shape(0); ++index) {
+
+      if (!include_halo and ghostView_(index)) continue;
```
It would be more efficient to hoist the `if(!include_halo)` out of the loop, so that you can avoid the 
`if(ghostView_(index))` in case you like halos

odlomax replied:
Since all the owned data should be at the beginning of the field, I've now kept track of the highest non-ghost index and iterated no further than that. Avoids if statements in the loop.

wdeconinck commented on src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc:
```diff
@@ -306,15 +316,13 @@ void CubedSphereMeshGenerator::generate_mesh( const CubedSphereGrid& csGrid, con
                               ( i > N && j > N ) ||  // Top-right corner.
                               ( i < 0 && j > N );    // Top-left corner.
 
-        if ( inCorner ) {
```
The clang-tidy would not approve removing the "readability braces"

odlomax replied:
I think that happened by accident when I merged branches. Whoops!

wdeconinck replied:
Understood, it's not just this instance btw.

odlomax replied:
Okie dokie. So in general: braces = good; no braces = bad.

wdeconinck reviewed (changes requested):
Good work!
Multiple inheritance is not really desired indeed...
You could do the following though to get desired behaviour, and perhaps more elegant?

```
template< typename BaseFunctionSpace >
class CubedSphereFunctionSpace : public BaseFunctionSpace {
public:
  StructuredCubedSphere( const Mesh& mesh ) :
    BaseFunctionSpace( mesh ) {
    ...
  }
  ...
};

using CubedSphereCellColumns = CubedSphereFunctionSpace<CellColumns>;
using CubedSphereNodeColumns = CubedSphereFunctionSpace<NodeColumns>;
```

MarekWlasak committed:
ATLAS-330 Setting execute_adjoint in interpolation.h to have a constant target. (#75)

odlomax committed:
Update src/atlas/functionspace/detail/StructuredCubedSphere.h

Co-authored-by: Willem Deconinck <wdeconinck@me.com>

wdeconinck commented on src/atlas/functionspace/detail/StructuredCubedSphere.h:
```diff
@@ -0,0 +1,113 @@
+  Field tij_;
+
+  // ghost field.
+  Field ghost_;
```
Another problem I think you have here, is that you're extending a ObjectHandle class and not the implementation Object class. This has implications that when you e.g do:
```
CubedSphereCellColumns a,b;
b = a;
```
Then also the member data `tij_`, `ghost_`, `ijBounds_`, `tijToIdx_` gets copied instead.

odlomax replied:
I think I've avoided this now by adding an object handle for the CubedSphereColumnsImpl.

odlomax committed:
Replaced multiple inheritance with a second object handle.

odlomax committed:
Merge branch 'feature/cubed-sphere-structured-functionspace' of https://github.com/JCSDA-internal/atlas into feature/cubed-sphere-structured-functionspace

odlomax committed:
Addressed reviewer comments.

odlomax commented:
Hi @wdeconinck,

I've replaced the multiple inheritance with a `Object`-`ObjectHandle` class pair. Can you check I've used them correctly? I'm assuming they behave a bit like `std::shared_ptr`.

I've also added something very similar to the SFINAE `parallel_for` in `StructuredColumns`. It's witchcraft!

I've also added a `levels` method to `CellColumns`. I can probably find a workaround if this was omitted for a good reason.

odlomax committed:
Minor typos.

odlomax committed:
Refactored SFINAE macro into typedefs.

odlomax committed:
Refactored SFINAE macro into typedefs.

wdeconinck committed:
Fix compilation and test with ATLAS_BITS_LOCAL=64

odlomax committed:
Corrected comments.

odlomax committed:
Corrected comments.

wdeconinck committed:
Add caching for the Structured interpretation, avoiding recomputations

wdeconinck commented:
I added some more tests for copying and reinterpreting existing function spaces.
I noticed that the structure was being recomputed every time, so I created a caching mechanism.
Upon destruction of the mesh, the structure is being wiped from cache.

I would probably rename the "CubedSphereColumnsImpl" class to "CubedSphereStructure"

wdeconinck committed:
Fix atlas_test_cubedsphere_meshgen with ATLAS_BITS_LOCAL=64

odlomax committed:
Reactored CubedSphereColumnsImpl class to CubedSphereStructure.

odlomax committed:
Reactored CubedSphereColumnsImpl class to CubedSphereStructure.

odlomax commented:
@wdeconinck  I've refactored the class as you suggested.

@wdeconinck @MarekWlasak Are you both happy for me to submit this PR to `ecmwf/atlas/develop`?

wdeconinck reviewed (approved):
All good 👍

MarekWlasak reviewed (approved):
I am happy for this to go as an ecmwf PR.  Well done. There is a lot of good work here.

odlomax committed:
Merge branch 'ecmwf:develop' into feature/cubed-sphere-structured-functionspace

wdeconinck committed:
Merge branch 'develop' into feature/cubed-sphere-structured-functionspace

odlomax committed:
Increased the error tolerance on equality checks.

odlomax committed:
Increased the error tolerance on equality checks.

wdeconinck committed:
Update cache version

MarekWlasak commented:
@odlomax - why did n't you just pull this change from ecwmf develop?

odlomax commented:
This is the branch that was merged into ECMWF develop. I assumed that _not_ merging it into feature/cubed-sphere-base2 was an oversight.
